hyppo/independence/friedmanRafsky.py

```diff
@@ -0,0 +1,297 @@
+    spanning tree via Prim's algorithm
+    Parameters
+    ----------
+    data : ndarry
+        Weighted connection matrix
+    Returns
+    -------
+    MST_connections : list
+        List of pairs of nodes connected in final MST.
+    """
+
+    INF = 9999999
+
+    V = len(labels)
+
+    selected = np.zeros(len(labels))
+
+    no_edge = 0
+
+    selected[0] = True
+
+    MST_connections = []
+
+    while no_edge < V - 1:
+
+        minimum = INF
+        x = 0
+        y = 0
+        for i in range(V):
+            if selected[i]:
+                for j in range(V):
+                    if (not selected[j]) and G[i][j]:
+                        # not in selected and there is an edge
+                        if minimum > G[i][j]:
+                            minimum = G[i][j]
+                            x = i
+                            y = j
+        MST_connections.append([x, y])
+        selected[y] = True
+        no_edge += 1
+
+    return MST_connections
+
+
+@jit(nopython=True, cache=True)
+def MST(data, algorithm):
+
+    r"""
+    Helper function to read input data and calculate Euclidean distance
+    between each possible pair of points before finding MST.
+    Parameters
+    ----------
+    data : ndarry
+        Dataset such that each column corresponds to a point of data.
+    Returns
+    -------
+    MST_connections : list
+        List of pairs of nodes connected in final MST.
+    """
+    if algorithm == "Kruskal":
+
+        g = self.Graph(len(data[0] - 1))
+
+        for i in range(len(data[0])):
+            j = i + 1
+
+            while j <= (len(data[0]) - 1):
+                weight = np.linalg.norm(data[:, i] - data[:, j])
+                g.add_edge(i, j, weight)
+                j += 1
+
+        MST_connections = g.kruskal()
+
+    if algorithm == "Prim":
+
+        G = zeros((len(data[0]), len(data[0])))
+
+        for i in range(len(data[0])):
+            j = i + 1
+
+        while j <= (len(data[0]) - 1):
+            weight = np.linalg.norm(data[:, i] - data[:, j])
+            G[i][j] = weight
+            G[j][i] = weight
+            j += 1
+
+        MST_connections = prim(G)
+
+    return MST_connections
+
+
+@jit(nopython=True, cache=True)
+def num_runs(labels, MST_connections):
+
+    r"""
+    Helper function to determine number of independent
+    'runs' from MST connections
+    Parameters
+    ----------
+    labels, MST_connections : ndarry, list
+        Lables corresponding to respective classes of points and MST_connections
+        defining which pairs of points remain connect in final MST.
+    Returns
+    -------
+    run_count : int
+        Number of runs after severing all such edges with nodes of
+        differing class labels.
+    """
+
+    run_count = 1
+
+    for x in MST_connections:
+        if labels[x[0]] != labels[x[1]]:
+            run_count += 1
+
+    return run_count
+
+
+@jit(nopython=True, cache=True)
+def permutation(perm, labels, MST_connections):

```

Comment:
what is this doing that `perm_test` does not?

Reply:
It was my understanding that perm_test rearranged the y values corresponding to given x values while all labels remained constant (i.e. a permuted x value still has an x label, y value still has a y label). This instead permutes the labels and then determines the new run count from the MST based on the same labels to calc the test statistic. If I'm misinterpreting the usage here just let me know

Reply:
What do you mean by labels?

For independence tests, `perm_test` computes the p-values by permuting the values of matrix `y` (which can contain labels) and recomputing the test statistic.

Reply:
The labels would be some value (0) for x points and some value (1) for y points. I guess I was misinterpreting the perm_test code. How would I incorporate the existing perm_test function into the code

Reply:
what you described is equivalent to the `perm_test` code minus the labels matrix. Permuting the labels like this is equivalent to permuting `y`.

Reply:
perm_test takes in data x such that each row is a data point it looks like while I've written the code to operate as if each column was a data point. Do you want me to rewrite the test itself to work for some x and y where rows of x are the data points of the combined samples and y is a column vector of the labels?

Reply:
Also to compute the test statistic for the true case I need to permute the labels to compare to the random run counts so I think I would have to permute anyways

Reply:
> perm_test takes in data x such that each row is a data point it looks like while I've written the code to operate as if each column was a data point. Do you want me to rewrite the test itself to work for some x and y where rows of x are the data points of the combined samples and y is a column vector of the labels?

just transpose the inputs

Reply:
> Also to compute the test statistic for the true case I need to permute the labels to compare to the random run counts so I think I would have to permute anyways

you are. What I am saying is that both ways are equivalent and valid ways to permute a dataset